feat: allow to enforce org/env on console/portal

guillaumelamirand opened the pull request:
## Description

Re-introduce the capability to enforce the organization or environment used by a console or a portal.
The enforced id is either : 

- extract from the previous base URL configured on the constants.json or config.json
- get from a new attribute in the constants.json or config.json

Helm chart has been updated in order to propose to define this enforced id.

In order to support both shared instances (multitenant) or multiple instance, a new installation type attribute has been introduced in the communication with Cockpit.

## Additional context

<!-- Add any other context about the PR here -->
<!-- It can be links to other PRs or docs or drawing -->
<!-- Or reproduction steps in case of bug fix -->

<!-- Environment placeholder -->

🏗️ Your changes can be tested here and will be available soon:
      Console: [https://pr.team-apim.gravitee.dev/5657/console](https://pr.team-apim.gravitee.dev/5657/console)
      Portal: [https://pr.team-apim.gravitee.dev/5657/portal](https://pr.team-apim.gravitee.dev/5657/portal)
      Management-api: [https://pr.team-apim.gravitee.dev/5657/api/management](https://pr.team-apim.gravitee.dev/5657/api/management)
      Gateway v4: [https://pr.team-apim.gravitee.dev/5657](https://pr.team-apim.gravitee.dev/5657)
      Gateway v3: [https://pr.gateway-v3.team-apim.gravitee.dev/5657](https://pr.gateway-v3.team-apim.gravitee.dev/5657)

<!-- Environment placeholder end -->
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-ltkgrcsfmt.chromatic.com)
<!-- Storybook placeholder end -->

Commit:
feat: allow to enforce org/env on console/portal

jgiovaresco commented on gravitee-apim-console-webui/src/index.ts:
```diff
@@ -86,6 +94,24 @@ function sanitizeBaseURLs(constants: any): any {
   return baseURL;
 }
 
+function getEnforcedOrganizationId(constants: any): any {
```
```suggestion
function getEnforcedOrganizationId(constants: any): string | undefined {
```

jgiovaresco commented on gravitee-apim-portal-webui/src/app/services/configuration.service.ts:
```diff
@@ -81,6 +87,24 @@ export class ConfigurationService {
     return baseURL;
   }
 
+  _getEnforcedEnvironmentId(config: any): any {
```
```suggestion
  _getEnforcedEnvironmentId(config: any): string | undefined {
```

Commit:
feat: add back organization in the v2 API

   - organization id has been introduced on v2 API to make it easier to override without access points
   - following path are now available :
      - /apis where organization and environment will be DEFAULT
      - /environments/{envId} where organization will be DEFAULT
      - /organizations/{orgId}/environements/{envId} where both must be specified
   - open api has been splitted into 2 files:
      - existing files contains only paths with orga and env; Different servers are available to make it easier to use
      - plugins file contains only /plugins paths
   - index.html now contains 2 links to those new files

Commit:
feat: define access point template on installation

   - introduce notion of installation access service in order to retrieve ui api from either access point or configuration depending on the installation type (multi-tenant or standalone)
   - update cockpit communication to handle different additional info depending on the type

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/index-management.html:
```diff
@@ -0,0 +1,20 @@
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+  <title>Management API v2 - API Docs</title>
```
```suggestion
  <title>Management API v2 - APIs</title>
```

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/index-plugins.html:
```diff
@@ -0,0 +1,20 @@
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+  <title>Management API v2 - API Docs</title>
```
```suggestion
  <title>Management API v2 - Plugins</title>
```

phiz71 reviewed (commented):
Just a little comment. I've not read everything ^^

guillaumelamirand commented:
It seems to be an issue with code coverage because I have created unit tests but the coverage doesn't seem to be accurate.

phiz71 commented:
> It seems to be an issue with code coverage because I have created unit tests but the coverage doesn't seem to be accurate.

Yes, it's due to the split of tests we did during the last tech sprint.
It's planned to fix that in the next tech sprint.

phiz71 commented on helm/templates/ui/ui-configmap.yaml:
```diff
@@ -28,11 +28,14 @@ data:
+      "baseURL": "{{ .Values.ui.baseURL }}",
       {{- else }}
-      "baseURL": "{{ .Values.api.ingress.management.scheme }}://{{ index .Values.api.ingress.management.hosts 0 }}{{ .Values.api.ingress.management.path }}"
+      "baseURL": "{{ .Values.api.ingress.management.scheme }}://{{ index .Values.api.ingress.management.hosts 0 }}{{ .Values.api.ingress.management.path }}",
```
I think you will have a formatting issue with these `,`.
Because if you set toe company name wihtout the enforceOrgId and the managementTitle you will end with something like this.

```
"baseURL": "{{ .Values.ui.baseURL }}",,
"company": {
  "name": .....
}
```

Maybe I'm wrong, I think some unit tests could be helpful here

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/installation/EnvironmentResource.java:
```diff
@@ -0,0 +1,57 @@
+        return EnvironmentMapper.INSTANCE.map(environmentService.findByOrgAndIdOrHrid(GraviteeContext.getCurrentOrganization(), envId));
+    }
+
+    @Path("/apis")
```
is it required since /apis is already defined in ApisResource ?

guillaumelamirand replied:
it is to allow /apis or /environments/id/apis or /organization/id/environment/id/apis

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/installation/OrganizationResource.java:
```diff
@@ -44,4 +44,9 @@ public class OrganizationResource extends AbstractResource {
         return OrganizationMapper.INSTANCE.map(organizationService.findById(orgId));
     }
+
+    @Path("/environments")
```
same question, is it required, since it's already defined in EnvironmentsResource ?

guillaumelamirand replied:
it is to allow /environments/id or /organization/id/environments/id

phiz71 commented on apim-result.xml:
Can be deleted I think

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/OpenAPIResource.java:
```diff
@@ -31,12 +31,33 @@ public class OpenAPIResource {
+    }
+
+    @GET
+    @Path("/openapi-plugins.yaml")
```
I think you forgot `openapi-ui`

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/OpenAPIResource.java:
```diff
@@ -31,12 +31,33 @@ public class OpenAPIResource {
+    }
+
+    @GET
+    @Path("/index-plugins.html")
```
and `index-ui.html`

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/index-ui.html:
```diff
@@ -0,0 +1,20 @@
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+  <title>Management API v2 - APIs</title>
```
```suggestion
  <title>Management API v2 - UI</title>
```

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml:
```diff
@@ -16,7 +16,13 @@ security:
     - url: /management/v2
-      description: Gravitee.io APIM - Management API -v2
+      description: Gravitee.io APIM - Management API - v2
+    - url: /management/v2/organizations/{orgId}
```
can't we add a third one ?

`- url: /management/v2/organizations/{orgId}/environments/{envId}`

and remove the 'environnments' part of the paths ?

guillaumelamirand replied:
no, we cant because of /environments itself and /organizations

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-standalone/gravitee-apim-rest-api-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -11,6 +11,8 @@
+#console:
 ## Pendo Analysis Tool
 # To make our SaaS offering easier to use, we now have the ability to track activity in Pendo. This is disabled by
 # default, but if you'd like to help us in this journey, don't hesitate to reach us to get our key ;)
```
duplicated line

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-standalone/gravitee-apim-rest-api-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -576,11 +576,63 @@ notifiers:
+#    # Specify the access points of your installation, mandatory if you want to connect it to Cockpit with a multi-tenant installation
+#    # You can use template variable such as {account}, {organization} or {environment}
+#    accessPoints:
+#      organization:
```
could we have a gateway in the organization ? Because today it's possible to link a gateway to a whole org.

guillaumelamirand replied:
you can link a gateway to on orga? Where can you do that? 
Regarding the feature itself the gateway URL must be environment related not organization related to allow restrictedDomains

phiz71 replied:
https://github.com/gravitee-io/gravitee-api-management/blob/master/gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-distribution/src/main/resources/config/gravitee.yml#L541

guillaumelamirand replied:
Everything here is only used by APIM, not the gateway. If you have defined a orga or an env in the gateway config, the gateway will filter the API based on that, nothing else.

phiz71 replied:
Thanks to our discussion, I finally understood 😅
It's ok 😉

phiz71 commented on helm/values.yaml:
```diff
@@ -1258,7 +1258,7 @@ portal:
     enabled: true
     pathType: Prefix
-    path: /
+    path: /portal(/.*)?
```
Coudl it be considered as a breaking change ? If yes, it should be documented (in the changelog.md ?)

guillaumelamirand replied:
actually, if you keep / it will never work...so I cannot understand how it work like that. Maybe you know more about it ?

sebdevaux replied:
I don't understand, actually if you use / it's works (eg: https://apim-4-1-x-portal.team-apim.gravitee.dev/). If you wan't to change this value, you have to add a rewrite target: nginx.ingress.kubernetes.io/rewrite-target: /$1 just like we do for /console

guillaumelamirand replied:
if you use the values.yml like that it doesn't because / is already used by the gateway.

For master it does because we have set a dedicated host for the portal (different than others hosts) see bellow. 
![image](https://github.com/gravitee-io/gravitee-api-management/assets/4324887/5f44b318-294f-4b1c-89f5-e9798e491284)

Actually, I don't want to change anything, just making work by default . I revert it doesn't matter for my PR.

phiz71 commented on pom.xml:
```diff
@@ -55,7 +55,7 @@
         <gravitee-bom.version>6.0.15</gravitee-bom.version>
         <gravitee-alert-api.version>1.9.1</gravitee-alert-api.version>
-        <gravitee-cockpit-api.version>2.2.0</gravitee-cockpit-api.version>
+        <gravitee-cockpit-api.version>2.3.0-feat-multi-tenancy-installation-type-SNAPSHOT</gravitee-cockpit-api.version>
```
Don't forget to validate this first ;)

phiz71 reviewed (commented):
LGTM. Just a lots of question for you ^^

ThibaudAV reviewed (commented):
wouldn't it be possible to do without this `baseURL` and have a configuration like :
```
{
  managementAPI :...
  managementAPIv2 : ..
  defaultOrganisationId: ..
}
```
And then it's up to the app to reconsider the url and define what the parameters are

jourdiw commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiResource.java:
Where is this class given a `@Path` ?

guillaumelamirand replied:
It has always been its parent > ApisResource.

jourdiw commented on gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/management-openapi-v2.yaml:
Could we show clearly in the open api doc that certain resources can be scoped in multiple ways? Such as --> `/apis or /environments/id/apis or /organization/id/environment/id/apis`

guillaumelamirand replied:
It is the case as you can define the server at the top.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/query_service/cockpit/CockpitAccessServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+ */
+package io.gravitee.apim.infra.query_service.cockpit;/*
+ * Copyright (C) 2015 The Gravitee team (http://gravitee.io)&#10;&#10;Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);&#10;you may not use this file except in compliance with the License.&#10;You may obtain a copy of the License at&#10;&#10;        http://www.apache.org/licenses/LICENSE-2.0&#10;&#10;Unless required by applicable law or agreed to in writing, software&#10;d ...
+ */
```
?

Commit:
fix: properly defined v2 api resources

Commit:
chore: bump cockpit-api version
